src/qutip_tensorflow/core/data/dense_tensor.py

```diff
@@ -0,0 +1,85 @@
+import qutip as qt
+import numpy as np
+import tensorflow as tf
+import qutip.core.data as data
+import numbers
+
+__all__ = ['DenseTensor']
+
+
+class DenseTensor(data.Data):
+    """This class provide a wraps around TensorFlow's Tensor.
+
+    Parameters
+    ----------
+    data: array-like
+        Data to be stored.
+    shape: (int, int)
+        Shape of data. Default None, it tries to infer the shape from data
+        accessing the attribute `data.shape`.
+    copy: bool
+        Default True. If True creates a copy of the data. If a `tf.Tensor` is
+        provided as data, it will only preserve the graph structure of the
+        original data if copy=False. If anything different to a `tf.Tensor` or
+        `tf.Variable` is provided, it will copy the data regardless of copy as
+        it needs to be moved to the GPU.
+    """
+
+    def __init__(self, data, shape=None, copy=False):
+        # Try to inherit shape from data
+        if shape is None:
+            try:
+                shape = data.shape
+            except AttributeError:
+                raise ValueError("""Shape could not be inferred from data.
+                                 Please, include the shape of data.""")
```

Comment:
It's not clear to me what inputs you're prepared to accept.  The docstrings says `array_like`, but this check will cause `DenseTensor([[1, 0], [0, 1]])` to raise an exception, despite it being a valid Numpy array-like object.  You might want to rethink the order you do things, so you're not repeating checks or throwing away valid input from which the shape _can_ be inferred.  Remember that the `__init__` method is really meant for _humans_ more than anything else - library code often uses fast alternative constructor methods, because it often already has a lot of knowledge about the data being passed in, and having to write more doesn't matter very much in library code.